Allow inlining methods with unmatched type parameters

ianatol opened the pull request:
Rebase of #44656 to solve some inference regressions in #44803

As Keno predicted, there are some regressions in this PR as well, including quite a few test suite failures that I will be fixing up.

Additionally, some needed future work would be to teach SROA how to fold `svec_ref` more generally, as it is currently very specific

The original description follows below.

---

Currently we do not allow inlining any methods that have unmatched
type parameters. The original reason for this restriction is that
I didn't really know what to put for an inlined :static_parameter,
so I just had inlining bail. As a result, in code like:

```
f(x) = Val{x}()
```

the call to `Val{x}()` would not be inlined unless `x` was known
through constant propagation.

This PR attempts to remidy that. A new builtin is added that computes
the static parameters for a given method/argument list. Additionally,
sroa gains the ability to simplify and fold this builtin. As a result,
inlining can insert an expression that computes the correct values
for the inlinees static parameters.

The change benchmarks favorably:

Before:
```
julia> function foo()
          for i = 1:10000
              Base.donotdelete(Val{i}())
          end
       end
foo (generic function with 1 method)

julia> @time foo()
  0.375567 seconds (4.24 M allocations: 274.440 MiB, 14.67% gc time, 72.96% compilation time)

julia> @time foo()
  0.012387 seconds (9.49 k allocations: 148.266 KiB)
```

After:
```
julia> function foo()
          for i = 1:10000
              Base.donotdelete(Val{i}())
          end
       end
foo (generic function with 1 method)

julia> @time foo()
  0.003058 seconds (29.47 k allocations: 1.546 MiB)

julia> @time foo()
  0.001200 seconds (9.49 k allocations: 148.266 KiB)
```

Note that this particular benchmark could also be fixed by #44654,
but this change is more general.

There is a potential downside, which is that we remove a specialization
barrier here. We already do that in the case when all type parameters
are matched, so it's not eggregious. However, there is anectodal
knowledge in the community that extra type parameters force specialization.
Some of that is due to the handling of type parameters in the specialization
code, but some of it might also be due to inlining's prior refusal
to perform this inlining. We'll have to keep an eye out for any
regressions.

ianatol commented:
Re: partial reverting of #44512:

[`choltype`](https://github.com/JuliaLang/julia/blob/master/stdlib/LinearAlgebra/src/cholesky.jl#L181) from `LinearAlgebra` was causing BoundsErrors during test suite runs.

Looking further, it appeared that in the following IR,

```julia
Expr(:call, Main.eltype, Core.Argument(n=2)),
  Expr(:call, Main.oneunit, SSAValue(1)),
  Expr(:call, Main.sqrt, SSAValue(2)),
  Expr(:call, Main.typeof, SSAValue(3)),
  Expr(:call, Main.promote_type, SSAValue(4), Main.Float32),
  Core.ReturnNode(val=SSAValue(5))]
```

we were choosing to inline `promote_type` with the following match: `promote_type(DataType, Type{Float32}) from promote_type(Type{T}, Type{T}) where {T}`. 

This triggered `_compute_sparams` where
`m = promote_type(Type{T}, Type{T}) where {T}`
`tt = Tuple{typeof(Base.promote_type), Type{Float64}, Type{Float32}}`, which resulted in `env = svec()`, and our BoundsError.

Keno helped me to figure out that this was due to inlining not being able to discern which signature was right due to the `spec_types` of `promote_type(Type{T}, Type{T}) where {T}` and `promote_type(Type{T}, Type{S}) where {T,S}` being identical in our case. However, `DataType` in `spec_types` would later become `Float64`, meaning that this selection was actually incorrect. 

By bringing back [this branch](https://github.com/JuliaLang/julia/pull/45062/commits/b2ea6d114bfe7516bf40bd7e557e0e171c873864#diff-c08d0dffd36a142536203edece93dbe8656a68c607fd06256903d4b2ed8f2942R1284-R1310) that was removed in #44512, we were able to be a bit more fine-grained about when we could bypass validating sparams (as well as choosing not to inline when we had a direct match of `sig` and `spec_types`).

ianatol commented:
@nanosoldier `runtests(ALL, vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1111434923) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c899fac_vs_ce5909c/report.html).

ianatol commented:
`GeometricFlux`, `GraphSignals`, `GridapEmbedded`, `GridapGmsh`, `JetPack`, `NeuralOperators`, `Org`, ` PDENLPModels`, and `ValueShapes` are all caused by the same bug in FillArrays where we end up with `compute_sparams` on

`Tuple{Type{FillArrays.Fill{T, N, Axes} where Axes where N where T}, T, Vararg{Integer, N}} where N where T`
and
`Tuple{Type{FillArrays.Fill{T, N, Axes} where Axes where N where T}, Float32, Tuple{Int64}}`

Seems similar to the case listed above, but still investigating where a possibly incorrect inlining choice happens

Commit:
Allow inlining methods with unmatched type parameters

ianatol commented:
@nanosoldier `runtests(["Evolutionary", "GeometricFlux", "GraphSignals", "GridapEmbedded", "GridapGmsh", "Hecke", "InformationGeometry", "JetPack", "Metida", "NeuralOperators", "ODEInterface", "ODEInterfaceDiffEq", "Org", "PDENLPModels", "ValueShapes"], vs = ":master")`

ianatol commented:
@nanosoldier `runtests(["Hecke", "Metida", "ODEInterface", "Org"], vs = ":master")`

ianatol commented:
@aviatesk Could I ask for your review here? Specifically on [3a6b36c](https://github.com/JuliaLang/julia/pull/45062/commits/3a6b36c062c36c3250ee299cf7a3ef6672262782) where I revert some of #44512, as well as on [884e5ea](https://github.com/JuliaLang/julia/pull/45062/commits/884e5eaea0443f216b26bfe426b3a4121edca5a5), where I somewhat broadly restrict the cases in which we can perform this inlining. It feels like we could do better, but I'm not familiar enough with this part of the codebase to fully understand

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1118041430) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/884e5ea_vs_862018b/report.html).

ianatol commented:
Re: test failures in `Metida` and `ODEInterface`:

The `ccall` and `cfunction`s showing up in the test aren't having their types properly calculated during inlining since we are entering `ssa_substitute_op!` while `spvals::SSAValue`. 

My idea is that we can do something similar to `svec_ref`, where we insert a node (in this case a call to `jl_instantiate_type_in_env`) where `spvals` is still an `SSAValue`. Then, before emitting the substituted `ccall`, we replace `spvals` in our inserted nodes with our calculation of the intersection.

aviatesk commented:
> @aviatesk Could I ask for your review here? Specifically on [3a6b36c](https://github.com/JuliaLang/julia/pull/45062/commits/3a6b36c062c36c3250ee299cf7a3ef6672262782) where I revert some of #44512, as well as on [884e5ea](https://github.com/JuliaLang/julia/pull/45062/commits/884e5eaea0443f216b26bfe426b3a4121edca5a5), where I somewhat broadly restrict the cases in which we can perform this inlining. It feels like we could do better, but I'm not familiar enough with this part of the codebase to fully understand

Sure, I will look into this later today or the weekend.

aviatesk commented on base/compiler/ssair/ir.jl:
```diff
@@ -1469,62 +1466,104 @@ function maybe_erase_unused!(
         end
-        return val
+    elseif isa(stmt, OldSSAValue)
+        return FixedNode(compact.ssa_rename[stmt.id], false)
```
Don't we need to fix the usage count here?
```suggestion
        val = compact.ssa_rename[stmt.id]
        if isa(val, SSAValue)
            compact.used_ssas[val.id] += 1
        end
        return FixedNode(compact.ssa_rename[stmt.id], false)
```

And if so, I think this part of `fixup_node` and `fixup_phinode_values!` should share the same code.

aviatesk commented on base/compiler/ssair/ir.jl:
```diff
@@ -1469,62 +1466,104 @@ function maybe_erase_unused!(
-    append!(compact.used_ssas, compact.new_new_used_ssas)
-    empty!(compact.new_new_used_ssas)
-    for idx in compact.late_fixup
+function just_fixup!(compact::IncrementalCompact, new_new_nodes_offset::Union{Int, Nothing} = nothing, late_fixup_offset::Union{Int, Nothing}=nothing)
```
```suggestion
function just_fixup!(compact::IncrementalCompact, new_new_nodes_offset::Union{Int, Nothing} = nothing, late_fixup_offset::Union{Int, Nothing} = nothing)
```

aviatesk commented on base/compiler/ssair/ir.jl:
```diff
@@ -1469,62 +1466,104 @@ function maybe_erase_unused!(
         values[i] = val
     end
-    values
+    return FixedNode(values, needs_fixup)
```
A usage of `FixedNode` here seems weird conceptually. It would be more natural to just return a tuple here.

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1280,19 +1351,30 @@ function handle_const_call!(
             j += 1
             result = results[j]
             any_fully_covered |= match.fully_covers
+            check_sparams = isa(only_method, Bool)
```
Isn't `only_method` being computed within this loop and so it's not fixed at this point?
I think we can use that after we exit this loop.

ianatol replied:
I think it is an OK usage here in terms of whether `meth` itself has >1 match or not, which should be all we care about for `validate_sparams` within `handle_match`

aviatesk commented on base/compiler/ssair/passes.jl:
```diff
@@ -702,6 +702,86 @@ function perform_lifting!(compact::IncrementalCompact,
+    val = stmt.args[4]
+    valT = argextype(val, compact)
+    (isa(valT, Const) && isa(valT.val, Int)) || return
+    valI = valT.val
```
```suggestion
    valI = valT.val::Int
```

aviatesk commented:
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

aviatesk commented on test/compiler/inline.jl:
```diff
@@ -1259,3 +1259,20 @@ end
     return maybe_error_int(1)
 end
+
+# basic tests for inlining of `apply_type` in the presence of unmatched type parameters
+f44656(x) = Val{x}()
+
+function g44656()
+    for i = 1:10000
+        Base.donotdelete(Val{i}())
+    end
+end
+
+let srcs = (code_typed1(f44656, (Any,)),
+            code_typed1(g44656))
+    for src in srcs
+        @test count(isnew, src.code) == 1
+        @test count(iscall((src, Core.apply_type), src.code)) == 0
+    end
+end
```
We should add more test cases. E.g. since this PR adds a change on the inlining pass for constant prop'ed callsite, we should have a corresponding test case.

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -359,11 +359,28 @@ function ir_inline_item!(compact::IncrementalCompact, idx::Int, argexprs::Vector
+            tuple_call = Expr(:call, TOP_TUPLE, def, nonva_args...)
+            tuple_type = tuple_tfunc(Any[argextype(arg, compact) for arg in nonva_args])
+            tupl = insert_node_here!(compact, NewInstruction(tuple_call, tuple_type, topline))
+            apply_iter_expr = Expr(:call, Core._apply_iterate, iterate, Core._compute_sparams, tupl, va_arg)
```
Don't we need to pass `def::Method`? And we also should have a test case that covers this pass.

ianatol replied:
We pass `def` in through `tupl`. I agree about a test case, though

aviatesk replied:
Ops, I missed it, never mind.

aviatesk commented:
> By bringing back [this branch](https://github.com/JuliaLang/julia/pull/45062/commits/b2ea6d114bfe7516bf40bd7e557e0e171c873864#diff-c08d0dffd36a142536203edece93dbe8656a68c607fd06256903d4b2ed8f2942R1284-R1310) that was removed in #44512, we were able to be a bit more fine-grained about when we could bypass validating sparams (as well as choosing not to inline when we had a direct match of `sig` and `spec_types`).

#44512 removed the previous (somewhat random imho) special casings, and you can recover some if needed. But I don't want to revert it this way, i.e. which optimizable cases handled by #44512 are now ignored, and which are still optimized? I think you essentially need to compute `only_method` specifically, and it can co-exist with the code after #44512.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1119411463) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/0c08729_vs_902a5c1/report.md).

ianatol commented:
Seems to be passing CI as well as locally passing previously problematic package tests (`Metida`, `Hecke`, etc).

I'd still like to do better on [4934be](https://github.com/JuliaLang/julia/pull/45062/commits/4934be9eb87829ea6d3cdae1ad1b5a952b820bbc)'s implementation, as well as add some more tests, but this could use PkgEval again as there have been a few changes since the last one.

@nanosoldier `runtests(ALL, vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1183787515) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/e50ba3a_vs_415c668/report.html).

ianatol commented:
@nanosoldier `runtests(["Alicorn", "AlphaStableDistributions", "AprilTags", "AstroChemistry", "DistributedArrays", "EBayes", "EnsembleKalmanProcesses", "Evolutionary", "GadgetIO", "GasChromatographySimulator", "Hashpipe", "HetaSimulator", "ITensorTDVP", "JumpProcesses", "LibExpat", "ManifoldDiffEq", "NumericalAlgorithms", "OpenTelemetryProto", "OptimKit", "OptimizationMetaheuristics", "Oscar", "Reproject", "Retriever", "SimplexGridFactory", "SymbolicUtils"], vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1184575641) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/d2a10aa_vs_67cee4b/report.html).

ianatol commented:
Still need to add more tests, but could I get another review @aviatesk?

aviatesk commented on base/compiler/ssair/ir.jl:
```diff
@@ -1490,63 +1487,88 @@ function maybe_erase_unused!(
+        return FixedNode(val, false)
     else
         urs = userefs(stmt)
+        needs_fixup = false
```
Doesn't this variable need to be updated for a case when `val isa NewSSAValue && reify_new_nodes === false`?

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1321,26 +1418,30 @@ end
 function handle_match!(
     match::MethodMatch, argtypes::Vector{Any}, flag::UInt8, state::InliningState,
-    cases::Vector{InliningCase}, allow_abstract::Bool = false)
+    cases::Vector{InliningCase}, allow_abstract::Bool = false, check_sparams::Bool = false)
```
Now it seems better to change the `allow_abstract` and `check_sparams` to required positional arguments or even required keyword ones, and make their configuration explicit at callsites (if they're positional, we should use `#=allow_abstract=#` inline comment also). It would make future refactoring easier and safer.

aviatesk replied:
(Although now I want to make it required argument, ) the default value of `check_sparams` should be `true` to keep the inlining algorithm safe.

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1321,26 +1418,30 @@ end
-    # processing this dispatch candidate
-    _any(case->case.sig === spec_types, cases) && return true
-    item = analyze_method!(match, argtypes, flag, state)
+    if check_sparams
+        # we may see duplicated dispatch signatures here when a signature gets widened
+        # during abstract interpretation: for the purpose of inlining, we can just skip
+        # processing this dispatch candidate
+        _any(case->case.sig === spec_types, cases) && return true
+        item = analyze_method!(match, argtypes, flag, state, true)
+    else
+        item = analyze_method!(match, argtypes, flag, state)
+    end
```
Why not
```suggestion
    # we may see duplicated dispatch signatures here when a signature gets widened
    # during abstract interpretation: for the purpose of inlining, we can just skip
    # processing this dispatch candidate
    _any(case->case.sig === spec_types, cases) && return true
    item = analyze_method!(match, argtypes, flag, state, check_sparams)
```
? Couldn't see any reason to skip the signature duplication check in the `else` branch.

ianatol replied:
I've refactored this, but in the case where there are unmatched type parameters, we need to further analyze since `spec_types` matching may not tell the whole story: see [this comment](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1110447843), though the other special casing ended up not being needed

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1321,26 +1418,30 @@ end
     spec_types = mi.specTypes
     allow_abstract || isdispatchtuple(spec_types) || return false
-    validate_sparams(mi.sparam_vals) || return false
+    check_sparams && (validate_sparams(mi.sparam_vals) || return false)
```
```suggestion
    check_sparams && !validate_sparams(mi.sparam_vals) && return false
```

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -873,8 +891,22 @@ function validate_sparams(sparams::SimpleVector)
+    src = m.source
+    isa(src, Vector{UInt8}) && (src = uncompressed_ir(m))
+    if isa(src, CodeInfo)
+        return src.has_fcall == true
```
```suggestion
        return src.has_fcall
```

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -873,8 +891,22 @@ function validate_sparams(sparams::SimpleVector)
     return true
 end
 
+function has_fcalls(m::Method)
```
Shouldn't this check be conservative and return `true` in unanalyzable case (and maybe should be renamed to `may_has_fcalls`)?

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1264,6 +1350,7 @@ function compute_inlining_cases(info::ConstCallInfo,
     local any_fully_covered = false
     local handled_all_cases = true
     local j = 0
+    local only_method = nothing # tri-valued: nothing if unknown, false if proven untrue, otherwise the method itself
```
This seems unused anymore.

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1321,26 +1418,30 @@ end
 function handle_const_prop_result!(
     result::ConstPropResult, argtypes::Vector{Any}, flag::UInt8, state::InliningState,
-    cases::Vector{InliningCase}, allow_abstract::Bool = false)
+    cases::Vector{InliningCase}, allow_abstract::Bool = false, check_sparams::Bool=false)
```
There seems to be no callsites of this function with `check_sparams` on.

ianatol replied:
I'd like to add on to this PR and allow unmatched type params for this kind of result in some cases, but for now I will remove this

aviatesk commented on base/compiler/ssair/passes.jl:
```diff
@@ -720,6 +720,86 @@ function perform_lifting!(compact::IncrementalCompact,
+    if isa(vec, SSAValue)
+        # TODO: We could do the whole lifing machinery here, but really all
+        # we want to do is clean this up when it got inserted by inlining,
+        # which always
```
```suggestion
        # which always targets simple `svec` call or `_compute_sparams`,
        # so this specialized lifting would be enough
```

ianatol commented:
Haven't been able to recreate the `SparseArrays/umfpack` error locally, so I guess it may be unrelated? Either way, this should be ready for another review when you get a chance! @aviatesk

aviatesk commented on src/builtins.c:
```diff
@@ -1613,6 +1613,36 @@ JL_CALLABLE(jl_f_finalizer)
+JL_CALLABLE(jl_f__svec_ref)
+{
+    JL_NARGS(_svec_ref, 3, 3);
+    jl_value_t *b = args[0];
```
Do we really need this first boundscheck argument? I'm not sure we will have codegen support for this builtin so that it can be used in a performance sensitive situation.

Keno replied:
I'd leave it - it's consistent with the other refs methods and gives us the option to add codegen if we want it.

aviatesk reviewed (commented):
The code looks good to me!

> As Keno predicted, there are some regressions in this PR as well

What's the status of this regression?

aviatesk commented on test/compiler/inline.jl:
```diff
@@ -1526,3 +1526,16 @@ function oc_capture_oc(z)
+    let f(x) = Val{x}()
+        g() = [ Base.donotdelete(Val{i}()) for i in 1:10000 ]
+        srcs = (code_typed1(f, (Any,)),
+                code_typed1(g))
```
```suggestion
                code_typed1(g, (Any,)))
```

ianatol replied:
This resulted in a test error, I think the initial way of calling it was correct

ianatol commented:
> What's the status of this regression? 

As far as I can tell, there is no longer a regression. I will try another PkgEval today, though the logic hasn't changed too much since the last one that came back clean (and I've been checking previously problematic packages locally). The regression Keno was talking about mostly manifested in some test suite failures, which are all resolved now.

Keno commented on src/ircode.c:
```diff
@@ -809,6 +809,8 @@ JL_DLLEXPORT jl_array_t *jl_compress_ir(jl_method_t *m, jl_code_info_t *code)
     JL_UNLOCK(&m->writelock); // Might GC
     JL_GC_POP();
+
+    write_uint8(s.s, code->has_fcall);
```
The stream here is already closed 10 lines earlier.

Keno commented on src/ircode.c:
```diff
@@ -892,6 +894,9 @@ JL_DLLEXPORT jl_code_info_t *jl_uncompress_ir(jl_method_t *m, jl_code_instance_t
         code->parent = metadata->def;
     }
+
+    code->has_fcall = read_uint8(s.s);
```
Same here

Keno commented on stdlib/Profile/src/Profile.jl:
```diff
@@ -521,7 +521,7 @@ function short_path(spath::Symbol, filenamecache::Dict{Symbol, String})
             end
             return path
-        elseif isfile(joinpath(Sys.BINDIR, Base.DATAROOTDIR, "julia", "src", "base", path))
+        elseif isfile(joinpath(Sys.BINDIR, Base.DATAROOTDIR, "julia", "base", path))
```
Rebase mistake?

Keno commented on base/compiler/ssair/passes.jl:
```diff
@@ -720,6 +720,87 @@ function perform_lifting!(compact::IncrementalCompact,
+            tvar = sig.var
+            sig = sig.body
+            isa(sig, DataType) || return
+            sig.name === Tuple.name
```
This does nothing.

ianatol replied:
I take it we want `sig.name === Tuple.name || return`?

Keno commented on stdlib/Serialization/src/Serialization.jl:
```diff
@@ -1191,6 +1191,9 @@ function deserialize(s::AbstractSerializer, ::Type{CodeInfo})
     end
     ci.propagate_inbounds = deserialize(s)
     ci.pure = deserialize(s)
+    if format_version(s) >= 19
```
I think you need to bump the format version to 20. Adding the field changes the serialization.

ianatol replied:
Done

Keno commented on test/compiler/inline.jl:
```diff
@@ -1394,9 +1394,7 @@ let src = code_typed1(Tuple{Any}) do x
-    # This requires more inlining enhancments. For now just make sure this
-    # doesn't error.
-    @test count(isnew, src.code) in (0, 1) # == 0
+    @test count(isnew, src.code) == 0
```
nice

ianatol replied:
❤️

Keno commented on test/operators.jl:
```diff
@@ -325,7 +325,7 @@ end
     @test sprint(show, Returns(1.0)) == "Returns{Float64}(1.0)"
 
-    illtype = Vector{Core._typevar(:T, Union{}, Any)}
+    illtype = Vector{Core.TypeVar(:T)}
```
What goes wrong here?

ianatol replied:
This change was just to use a cleaner `TypeVar` interface in this test, but the problem here was doing our special inlining with `compute_sparams` targeting a `Type` from the arguments, which created some problems with ambiguities. I disabled this kind of inlining if there is an argument `x` such that `x === Type && isdefined(x, :var)`

Keno reviewed (approved):
LGTM overall, could use a PkgEval run.

ianatol commented:
@nanosoldier `runtests(ALL, vs = ":master")`

Keno commented:
@maleadt is PkgEval running on this? It didn't show up in the status list.

ianatol commented:
@nanosoldier `runtests(ALL, vs = ":master")`

maleadt commented:
Maybe because your membership isn't public? Let's see if I trigger

@nanosoldier `runtests(ALL, vs = ":master")`

I just conveniently lost sudo access to the machine, so can't debug this.

ianatol commented:
My membership to JuliaLang and JuliaComputing are both public (though that definitely caught me in the past lol). It still doesn't seem to be running, so I guess there's some other issue?

maleadt commented:
> It still doesn't seem to be running, so I guess there's some other issue?

Yeah, the machine PkgEval is running on lost its DNS configuration. I think that killed Kerberos, locking me mostly out. So this may take a couple of days.

maleadt commented:
Should be working again:

@nanosoldier `runtests(ALL, vs = ":master")`

maleadt commented:
Sorry, restarting the job so that it uses a newer rootfs to build Julia (fixing certain package issues):

@nanosoldier `runtests(ALL, vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1222396964) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8619b4c_vs_e0056c8/report.html).

ianatol commented:
Hmm, seems we aren't detecting foreign calls to disable this kind of inlining correctly. These other failures I'm not as sure about. Hopefully they go away on a retry: 

@nanosoldier `runtests(["AbstractAlgebra", "BenchmarkFunctions", "BenchmarkProfiles", "Binscatters", "CairoMakie", "CalibrationAnalysis", "Compat", "CopEnt", "CountdownNumbers", "Enigma", "Evolutionary", "FMI", "FlameGraphs", "Folds", "ForwardDiff", "FunctionOperators", "ImageGeoms", "InformationInequalities", "KitePodModels", "MatrixProfile", "Mazes", "MultiplesOfPi", "NODAL", "NeuralDynamics", "ODEInterface", "ODEInterfaceDiffEq", "OptimKit", "Org", "PGENFiles", "ParameterEstimateScatterPlots", "Pidfile", "PlantGeom", "PlatformAware", "ProxSDP", "QuadEig", "QuadraticToBinary", "QuantumTomography", "ReinforcementLearningExperiments", "RetroCap", "RigidBodyTools", "RootedTrees", "RoundAndSwap", "SDMX", "SignalTablesInterface_WGLMakie", "SolverBenchmark", "SpatialEcology", "SteadyStateDiffEq", "StochasticProcesses", "StrBase", "TensorBoardLogger", "TensorGames", "Transits"], vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1223276957) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8619b4c_vs_80e50b5/report.html).

ianatol commented:
@nanosoldier `runtests(["CairoMakie", "Compat", "CountdownNumbers", "FlameGraphs", "Folds", "MultiplesOfPi", "ODEInterface", "ODEInterfaceDiffEq", "Org", "PGENFiles", "Pidfile", "PlatformAware", "QuantumTomography", "ReinforcementLearningExperiments", "RetroCap", "RoundAndSwap", "SteadyStateDiffEq", "StrBase", "TensorBoardLogger", "Transits"], vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1226073146) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/ecd6b27_vs_3b1c54d/report.html).

ianatol commented:
Of the few remaining PkgEval errors, I am unable to recreate them on multiple machines locally and they seem fairly unrelated. I think this is good to go.

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -875,8 +893,26 @@ function validate_sparams(sparams::SimpleVector)
+
+function can_inline_typevars(m::MethodMatch, argtypes::Vector{Any})
+    may_have_fcalls(m.method) && return false
+    any(x -> x === Type && isdefined(x, :var), argtypes[2:end]) && return false
```
```suggestion
    any(@nospecialize(x) -> x === Type && isdefined(x, :var), argtypes[2:end]) && return false
```
also, `x === Type && isdefined(x, :var)` can be simplified to `x === Type`.

aviatesk replied:
I guess you want `x isa UnionAll` instead?

aviatesk commented on base/compiler/ssair/passes.jl:
```diff
@@ -720,6 +720,93 @@ function perform_lifting!(compact::IncrementalCompact,
     return stmt_val # N.B. should never happen
 end
 
+has_typevar(@nospecialize(t), v::TypeVar) = ccall(:jl_has_typevar, Cint, (Any, Any), t, v)!=0
```
How about moving this to `compiler/utilities.jl`?

aviatesk commented on test/compiler/inline.jl:
```diff
@@ -1519,3 +1517,27 @@ function oc_capture_oc(z)
+                code_typed1(g),
+                code_typed1(h))
+        for src in srcs
+            @test !any(x -> isexpr(x, :call) && length(x.args) == 1, src.code)
```
```suggestion
            @test !any(@nospecialize(x) -> isexpr(x, :call) && length(x.args) == 1, src.code)
```

aviatesk commented on base/compiler/ssair/passes.jl:
```diff
@@ -720,6 +720,93 @@ function perform_lifting!(compact::IncrementalCompact,
+            sig = m.sig
+            isa(sig, UnionAll) || return
+            tvar = sig.var
+            sig = sig.body
+            isa(sig, DataType) || return
+            sig.name === Tuple.name || return
+            length(sig.parameters) >= 1 || return
+
+            i = findfirst(j->has_typevar(sig.parameters[j], tvar), 1:length(sig.parameters))
+            i === nothing && return
+            _any(j->has_typevar(sig.parameters[j], tvar), i+1:length(sig.parameters)) && return
+
+            arg = sig.parameters[i]
+            isa(arg, DataType) || return
+
+            rarg = def.args[2 + i]
+            isa(rarg, SSAValue) || return
+            argdef = compact[rarg][:inst]
+            if isexpr(argdef, :new)
+                rarg = argdef.args[1]
+                isa(rarg, SSAValue) || return
+                argdef = compact[rarg][:inst]
+            end
+
+            is_known_call(argdef, Core.apply_type, compact) || return
+            length(argdef.args) == 3 || return
+
+            applyT = argextype(argdef.args[2], compact)
+            isa(applyT, Const) || return
+            applyT = applyT.val
+
+            isa(applyT, UnionAll) || return
+            applyTvar = applyT.var
+            applyTbody = applyT.body
+
+            isa(applyTbody, DataType) || return
+            applyTbody.name == arg.name || return
+            length(applyTbody.parameters) == length(arg.parameters) == 1 || return
+            applyTbody.parameters[1] === applyTvar || return
+            arg.parameters[1] === tvar || return
```
Maybe we want to extract this part into a separate function?

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1226752107) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/1c0e83b_vs_e7ef4b5/report.md).

aviatesk commented:
Congrats, @ianatol !

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -376,11 +376,28 @@ function ir_inline_item!(compact::IncrementalCompact, idx::Int, argexprs::Vector
     end
+    if !validate_sparams(sparam_vals)
+        if def.isva
+            nonva_args = argexprs[1:end-1]
+            va_arg = argexprs[end]
+            tuple_call = Expr(:call, TOP_TUPLE, def, nonva_args...)
+            tuple_type = tuple_tfunc(Any[argextype(arg, compact) for arg in nonva_args])
+            tupl = insert_node_here!(compact, NewInstruction(tuple_call, tuple_type, topline))
+            apply_iter_expr = Expr(:call, Core._apply_iterate, iterate, Core._compute_sparams, tupl, va_arg)
+            sparam_vals = insert_node_here!(compact,
+                effect_free(NewInstruction(apply_iter_expr, SimpleVector, topline)))
```
this seems like it would add significant cost (and allocations and new DataType's) that were not present before inlining? are we sure this is profitable?

vtjnash replied:
Prior to the `fix_va_argexprs!` call that you have above, the argexprs list was required to be a known length and did not have or need a va_arg

Keno replied:
Addressed in https://github.com/JuliaLang/julia/pull/46700.

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1624,30 +1708,37 @@ function late_inline_special_case!(
                 ccall(:jl_instantiate_type_in_env, Any, (Any, Any, Ptr{Any}), argt, spsig, spvals)
                 for argt in e.args[4]::SimpleVector ]...)
-        elseif head === :foreigncall
+        elseif head === :foreigncall && isa(spvals, SimpleVector)
             @assert !isa(spsig, UnionAll) || !isempty(spvals)
```
shouldn't this be an assert? if it managed to get this far, but didn't run this fixup code here, it will generate corrupt code later and possibly segfault later at runtime

Keno replied:
I'm assuming you're talking about the `isa`? If so, I don't think so - there can be `:foreigncall`s that got inlined from a callee that don't need the fixup here (though it's not harmful either).

vtjnash commented on base/compiler/ssair/passes.jl:
```diff
@@ -720,6 +720,97 @@ function perform_lifting!(compact::IncrementalCompact,
     return stmt_val # N.B. should never happen
 end
 
+function lift_svec_ref!(compact::IncrementalCompact, idx::Int, stmt::Expr)
```
This should probably be removed and replaced with an inference `tfunc` for `_svec_ref`, since it is rather unfortunate (generally) when optimization ends up with a more precise answer than inference; it tends to confuse and annoy people, since it is usually less reliable and accurate.

Keno replied:
This lifts `svec_ref(_compute_sparams(...), ...)` which is inserted by inlining and does not exist at inference time.

vtjnash replied:
It seems possible that the optimizer could have already lifted that static_parameter expr into a `typeof` call inside the method, before it reached here. Is the optimizer not accurately marking the `svec_ref` with the result type constant (if known from inference) when it inserts it? Still seems confusing that inference couldn't determine this static_parameter value when the IR was simple, but optimization could figure it out after the IR was made more complicated.

Keno replied:
If the optimizer knows the type of the static parameter, it shouldn't insert the svec_ref (it currently does on master, but that's a bug fixed in https://github.com/JuliaLang/julia/pull/46703). Note that the optimizer here is not figuring out the *value* of the static parameters, but merely an SSAValue that will contain that value at runtime, which is a different analysis.

vtjnash replied:
Ah, okay, I thought this was attempting to forward the type. It makes more sense that this is attempting to lift the computation of `T` through a call that just gets the parameter back directly, e.g. `eltype(A{T}) -> T`

vtjnash commented on base/compiler/ssair/passes.jl:
```diff
@@ -720,6 +720,97 @@ function perform_lifting!(compact::IncrementalCompact,
+    m = m.val
+    isa(m, Method) || return nothing
+    # TODO: More general structural analysis of the intersection
+    length(def.args) >= 3 || return nothing
```
this must be earlier in the function, before you access `def.args[2]`

vtjnash commented on base/compiler/ssair/passes.jl:
```diff
@@ -720,6 +720,97 @@ function perform_lifting!(compact::IncrementalCompact,
+    sig.name === Tuple.name || return nothing
+    length(sig.parameters) >= 1 || return nothing
+
+    i = findfirst(j->has_typevar(sig.parameters[j], tvar), 1:length(sig.parameters))
+    i === nothing && return nothing
+    _any(j->has_typevar(sig.parameters[j], tvar), i+1:length(sig.parameters)) && return nothing
+
+    arg = sig.parameters[i]
+    isa(arg, DataType) || return nothing
+
+    rarg = def.args[2 + i]
+    isa(rarg, SSAValue) || return nothing
+    argdef = compact[rarg][:inst]
+    if isexpr(argdef, :new)
+        rarg = argdef.args[1]
+        isa(rarg, SSAValue) || return nothing
+        argdef = compact[rarg][:inst]
+    end
+
+    is_known_call(argdef, Core.apply_type, compact) || return nothing
+    length(argdef.args) == 3 || return nothing
+
+    applyT = argextype(argdef.args[2], compact)
+    isa(applyT, Const) || return nothing
+    applyT = applyT.val
+
+    isa(applyT, UnionAll) || return nothing
+    applyTvar = applyT.var
+    applyTbody = applyT.body
+
+    isa(applyTbody, DataType) || return nothing
+    applyTbody.name == arg.name || return nothing
+    length(applyTbody.parameters) == length(arg.parameters) == 1 || return nothing
+    applyTbody.parameters[1] === applyTvar || return nothing
+    arg.parameters[1] === tvar || return nothing
+    return argdef.args[3]
```
I need to think about this logic more, but this code needs substantial comment about the intent here. As written, I don't think the structural assumptions here are valid outside of trivial common cases (though it appears to try to limit itself to only the most trivial of trivial cases). But with so sparse comments, it is hard to review whether it is doing only what it intends and not including cases that it won't handle properly

vtjnash commented on base/compiler/ssair/passes.jl:
```diff
@@ -720,6 +720,97 @@ function perform_lifting!(compact::IncrementalCompact,
+            end
+            return
+        elseif is_known_call(def, Core._compute_sparams, compact)
+            res = _lift_svec_ref(def, compact)
```
how can you lift an svec_ref for `valI` without including the `valI` for the index of the value that you are referencing as an argument?

Keno replied:
It currently assumes `valI == 1` and that the svec is inbounds, but yeah, that should be checked (and generalized in the future).

vtjnash commented on base/compiler/ssair/passes.jl:
```diff
@@ -720,6 +720,97 @@ function perform_lifting!(compact::IncrementalCompact,
+
+    if isa(vec, SimpleVector)
+        if valI <= length(val)
+            compact[idx] = vec[valI]
```
This is missing a QuoteNode
